window/src/os/x11/keyboard.rs

```diff
@@ -397,6 +400,27 @@ impl KeyboardWithFallback {
             // Mod4
             res |= Modifiers::SUPER;
         }
+
+        // When XKB doesn't have modifiers in its state but there are modifiers given in the X
+        // event, use these as fallback.
+        //
+        // This can happen (FIXME: WHY) for example when Espanso (a text expander/injector)
+        // simulate inputs.
+        if let (true, Some(event_xmods)) = (res.is_empty(), fallback_xmods) {
```

Comment:
one thing I'm a little unsure about here is how to distinguish between:

* xkb state knew the modifier state but legitimately produced an empty set
* xkb has no idea and produced an empty set of mods

I'm not sure if this is purely a theoretical distinction or whether there is something more tricky we need to do in order to be sure about this.

Is there a return code we can use to influence this choice?

Reply:
or maybe it doesn't matter and we should unconditionally "or" the event-provided modifiers in?

Reply:
I don't think we can do this distinction, but I'd argue it doesn't matter much IMO.

Currently the PR uses the event mods as a fallback only, however I'm thinking it could be different:
1. when the X event is synthetic (was crafted and sent by another app), use event mods _instead_ of xkb detected mods
2. when the X event has mods, use event mods _instead_ of xkb detected mods
3. merge xkb mods and event mods (as you suggested)

Not sure what's the best direction here..
I think the behavior 1 can be a good idea for synthetic events, because if an app sends an xevent to another app it might want a specific set of mods, but at the same time it could also have sent events to indicate mods are pressed followed by the keypress/release event..

Maybe we could do 1 & 2 or 1 & 3, WDYT ?

It would be interesting to have more data about your issue, can you give the output of `xev -event keyboard` and pressing your paste key?

Reply:
Just dropping by to share the logging, haven't had a chance to ponder the questions so far.

```
KeyPress event, serial 28, synthetic NO, window 0x2e00001,
    root 0x6d1, subw 0x0, time 64899989, (300,145), root:(344,1595),
    state 0x0, keycode 50 (keysym 0xffe1, Shift_L), same_screen YES,
    XLookupString gives 0 bytes:
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x2e00001,
    root 0x6d1, subw 0x0, time 64899989, (300,145), root:(344,1595),
    state 0x1, keycode 118 (keysym 0xff63, Insert), same_screen YES,
    XLookupString gives 0 bytes:
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x2e00001,
    root 0x6d1, subw 0x0, time 64899990, (300,145), root:(344,1595),
    state 0x1, keycode 118 (keysym 0xff63, Insert), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x2e00001,
    root 0x6d1, subw 0x0, time 64899991, (300,145), root:(344,1595),
    state 0x1, keycode 50 (keysym 0xffe1, Shift_L), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False
```